src/profile-logic/import/art-trace.js

```diff
@@ -0,0 +1,932 @@
+  const magic = reader.getU32();
+  validateMagicHeader(magic);
+  const version = reader.getU16() ^ STREAMING_TRACE_VERSION_MASK;
+  validateVersion(version);
+  const headerSize = reader.getU16();
+  const startTimeInUsecSinceBoot = reader.getU64();
+  const recordSize = parseRecordSize(reader, version);
+  reader.setCurPos(headerStart + headerSize);
+
+  // Parse contents.
+  const methods = [];
+  const threads = [];
+  const recordStartPositions = [];
+  let summaryDetails = { clock: 'thread-cpu' };
+  while (!reader.eof()) {
+    const recordStart = reader.curPos();
+    const nonZeroRecordFragmentOrZero = reader.getU16();
+
+    if (nonZeroRecordFragmentOrZero !== 0) {
+      // A normal record. Store its position so that we can parse it later.
+      // We cannot parse it now because we need the clock information, which is
+      // located in the summary section at the end of the file.
+      recordStartPositions.push(recordStart);
+      reader.setCurPos(recordStart + recordSize);
+      continue;
+    }
+
+    // nonZeroRecordFragmentOrZero being 0 indicates a special action, which is
+    // identified by the code in the next byte.
+    const code = reader.getU8();
+    if (code === PARSE_METHOD) {
+      // Parse one method.
+      const methodStringLength = reader.getU16();
+      methods.push(parseOneMethod(reader.getString(methodStringLength)));
+    } else if (code === PARSE_THREAD) {
+      // Parse one thread.
+      const tid = reader.getU16();
+      const threadNameLength = reader.getU16();
+      const threadName = reader.getString(threadNameLength);
+      threads.push({ tid, threadName });
+    } else if (code === PARSE_SUMMARY) {
+      // Parse the entire summary section, and exit this loop.
+      const summaryLength = reader.getU32();
+      const summaryBytes = reader.getBytes(summaryLength);
+      const summary = parseSummary(new ByteReader(summaryBytes));
+      validateMatchingVersions(version, summary.summaryVersion);
+      summaryDetails = summary.summaryDetails;
+      break;
+    } else {
+      throw new Error(`Invalid trace format: got invalid code ${code}.`);
+    }
+  }
+
+  // Now that we have the "clock" information from the summary section, read the
+  // records from earlier in the file.
+  const { clock } = summaryDetails;
+  const methodActions = [];
+  for (const recordStart of recordStartPositions) {
+    reader.setCurPos(recordStart);
+    methodActions.push(parseRecord(reader, version, recordSize, clock));
+  }
+
+  return {
+    summaryDetails,
+    threads,
+    methods,
+    methodActions,
+    startTimeInUsecSinceBoot,
+  };
+}
+
+function parseArtTrace(buffer: ArrayBuffer): ArtTrace {
+  const reader = new ByteReader(new Uint8Array(buffer));
+  switch (detectArtTraceFormat(buffer)) {
+    case 'regular':
+      return parseRegularFormat(reader);
+    case 'streaming':
+      return parseStreamingFormat(reader);
+    default:
+      throw new Error('Not an android trace');
```

Comment:
Does this error get surfaced to the end-user? It might be better to be a bit more verbose here if so.

Reply:
This error should never be displayed because the parsing code is only entered if `isArtTraceFormat` returns `true` which it only does if `detectArtTraceFormat` returns `'regular'` or `'streaming'`. But the error re-wrapping commit that I added in response to your other comment would help with this case too.